### PR TITLE
Rewrite `editor-searchable-dropdowns` and add new options in dropdowns

### LINKED ISSUE #2753: [searchable-dropdowns]: add button to add new broadcasts and variables.
> make the search box be able to create messages and variables directly
image

To create variables, maybe there could be some sort of filtered thing, that also sets the variable's scope?

### PR DESCRIPTION
**Pull request by TheColaber**
Resolves #2753

hi

### CODE REVIEW

File: addons/editor-searchable-dropdowns/userscript.js [hunk 1]
```diff
@@ -1,61 +1,95 @@
...
     if (!hasScrollBar) {
-      const blocklyDropDownArrow = blocklyDropDownDiv.querySelector(".blocklyDropDownArrow");
-      if (blocklyDropDownArrow.classList.contains("arrowBottom")) {
+      const blocklyDropDownArrow = Blockly.DropDownDiv.arrow_;
+      if (!arrowAtTop) {
         const searchBarHeight = searchBar.offsetHeight;
         blocklyDropDownDiv.style.transform += ` translateY(-${searchBarHeight}px)`;
         blocklyDropDownArrow.style.transform = `translateY(${searchBarHeight}px) ${blocklyDropDownArrow.style.transform}`;
       }
     }
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:75]: remove all of this, no longer necessary and causes issues

File: addons/editor-searchable-dropdowns/userscript.js [hunk 2]
```diff
@@ -1,61 +1,95 @@
...
 
-  function createSearchBar(node) {
-    blocklyDropdownMenu = node;
+  let searchBar = null;
+  let fieldVariable = null;
+
+  const oldDropDownDivShow = Blockly.DropDownDiv.show;
+  Blockly.DropDownDiv.show = function (owner, ...args) {
+    fieldVariable = owner;
+    const arrowAtTop = oldDropDownDivShow.call(this, owner, ...args);
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:20]: you have to return this at the end

File: addons/editor-searchable-dropdowns/userscript.js [hunk 3]
```diff
@@ -1,61 +1,95 @@
...
-  }
+  };
 
-  function cleanup() {
+  const oldDropDownDivHide = Blockly.DropDownDiv.hide;
+  Blockly.DropDownDiv.hide = function () {
+    oldDropDownDivHide.call(this);
+    Blockly.DropDownDiv.content_.style.width = "";
+    Blockly.DropDownDiv.content_.style.height = "";
     blocklyDropdownMenu = null;
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:85]: do all of this in ScratchBlocks.DropDownDiv.clearContent instead

File: addons/editor-searchable-dropdowns/userscript.js [hunk 4]
```diff
@@ -1,61 +1,95 @@
...
 
     searchBar.type = "text";
     searchBar.addEventListener("input", handleInputEvent);
     searchBar.addEventListener("keydown", handleKeyDownEvent);
     searchBar.classList.add("u-dropdown-searchbar");
-    blocklyDropdownMenu.insertBefore(searchBar, blocklyDropdownMenu.firstChild);
+
+    const button = document.createElement("button");
+    button.addEventListener("click", () => {
+      const variable = Blockly.getMainWorkspace().createVariable(searchBar.value, "broadcast_msg");
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:43]: people should not be able to create broadcasts with empty names

File: addons/editor-searchable-dropdowns/userscript.js [hunk 5]
```diff
@@ -1,61 +1,95 @@
...
 
   function selectItem(item, click) {
-    // You can't just use click() or focus() because Blockly uses mousedown and mouseup handlers, not click handlers.
-    item.dispatchEvent(new MouseEvent("mousedown", { relatedTarget: item, bubbles: true }));
-    if (click) {
-      item.dispatchEvent(new MouseEvent("mouseup", { relatedTarget: item, bubbles: true }));
+    console.log(item);
+    if (click) fieldVariable.onItemSelected(null, item);
+    else {
+      // Hover somehow
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:92]: any reason you can't use fake mouse events, the way it used to?

**Review by GarboMuffin — CHANGES_REQUESTED** [review]

File: addons/editor-searchable-dropdowns/userscript.css [hunk 6]
```diff
@@ -11,6 +21,7 @@
   font-size: 13px;
   font-weight: bold;
   border-radius: 4px;
+  margin-right: 5px;
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.css:24]: margin-left on the button instead

**Review by GarboMuffin — CHANGES_REQUESTED** [review]

**Comment by GarboMuffin** [comment]
Also the dropdown div is used for not just dropdowns, but also direction and color pickers. Make sure those still work (they don't)

**Comment by GarboMuffin** [comment]
What if instead of a button by the search bar, you did something like:

image

Where that would always be at the bottom of a search. People would be able to use arrows and pressing enter to select it, as they would any other item. You could have separate "Create global variable Y" and "Create local variable X" items for variables.

Also, up and down arrows are currently broken

File: addons/editor-searchable-dropdowns/userscript.js [hunk 7]
```diff
@@ -1,62 +1,106 @@
...
+    Blockly.DropDownDiv.content_.style.height = "";
+  };
+
+  const oldFieldDropdownGetOptions = Blockly.FieldDropdown.prototype.getOptions;
+  Blockly.FieldDropdown.prototype.getOptions = function () {
+    fieldDropdown = this;
+    const options = oldFieldDropdownGetOptions.call(this);
+    const block = this.sourceBlock_;
+    if (block) {
+      console.log(block);
```
Comment (GarboMuffin) [inline addons/editor-searchable-dropdowns/userscript.js:61]: remove

File: addons-l10n/en/editor-searchable-dropdowns.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+  "editor-searchable-dropdowns/createGlobalVariable": "Create variable \"{name}\" for all sprites",
+  "editor-searchable-dropdowns/createLocalVariable": "Create variable \"{name}\" for this sprite only",
+  "editor-searchable-dropdowns/createGlobalList": "Create list \"{name}\" for all sprites",
+  "editor-searchable-dropdowns/createLocalList": "Create list \"{name}\" for this sprite only",
+  "editor-searchable-dropdowns/createBroadcast": "Create broadcast \"{name}\""
```
Comment (GarboMuffin) [inline addons-l10n/en/editor-searchable-dropdowns.json:6]: The closest string in Scratch is "New message", so perhaps "Create message"? Not sure.

**Review by GarboMuffin — CHANGES_REQUESTED** [review]
this is cool

More thoughts:

 - Should "new message" be hidden when something is searched?
 - Maybe rename and delete variable should be hidden as well?
 - See if "create variable" etc. can be hidden if nothing searched

**Comment by TheColaber** [comment]
> Maybe rename and delete variable should be hidden as well?

I had an idea to make these two items live update just like the others so you could rename the variable by just typing something in?

**Comment by WorldLanguages** [comment]
Sry for taking long whats the status of this? :P

**Comment by TheColaber** [comment]
I'm on vacation rn but I know this PR can
A. get merged - no problems at all
B. implement my suggestion above and then get merged

**Comment by GarboMuffin** [comment]
@TheColaber merge https://github.com/TheColaber/ScratchAddons/pull/24 and then I think this is done unless you plan to add more features

**Comment by WorldLanguages** [comment]
Can anything be done about the width of the dropdown?
image

**Comment by WorldLanguages** [comment]
AFAIK there's no easy solution. Unless we want the dropdown to grow as the user types more characters

**Review by WorldLanguages — APPROVED** [review]

**Pull request merged**